### PR TITLE
plan: add opt flag.

### PR DESCRIPTION
**Pull request by hanfei1991**
Not all queries need to use all the optmizer rules. This pr will make optimizer query-awared.
@shenli @coocood @zimulala @tiancaiamao @winoros @lamxTyler PTAL

### CODE REVIEW

File: plan/optimizer.go [hunk 1]
```diff
@@ -27,6 +27,28 @@ import (
...
+var optRuleList = []optimizeRule{
+	&decorrelateSolver{},
+	&ppdSolver{},
+	&columnPruner{},
+	&buildKeySolver{},
+	&aggPruner{},
+	&aggPushDownSolver{},
+}
+
+type optimizeRule interface {
```
Comment (shenli) [inline plan/optimizer.go]: Add comment for this.

File: plan/optimizer.go [hunk 2]
```diff
@@ -86,6 +109,7 @@ func doOptimize(logic LogicalPlan, ctx context.Context, allocator *idAllocator)
 	addCachePlan(pp, allocator)
 	log.Debugf("[PLAN] %s", ToString(pp))
 	return pp, nil
+
```
Comment (shenli) [inline plan/optimizer.go]: Useless line.

File: plan/optimizer.go [hunk 3]
```diff
@@ -45,37 +67,38 @@ func Optimize(ctx context.Context, node ast.Node, is infoschema.InfoSchema) (Pla
...
 		return nil, errors.Trace(ErrCartesianProductUnsupported)
 	}
+	logic.ResolveIndicesAndCorCols()
+	return physicalOptimize(logic, allocator)
+}
+
+func logicalOptimize(flag uint64, logic LogicalPlan, ctx context.Context, alloc *idAllocator) (LogicalPlan, error) {
+	var err error
+	for i, rule := range optRuleList {
+		if flag&(1<<uint(i)) == 0 {
```
Comment (shenli) [inline plan/optimizer.go]: A little tricky, use a better way or add some comments.

File: plan/decorrelate.go [hunk 4]
```diff
@@ -48,8 +49,10 @@ func (a *Apply) extractCorColumnsBySchema() {
 	a.corCols = resultCorCols[:length]
 }
 
+type decorrelateSolver struct{}
+
 // decorrelate function tries to convert apply plan to join plan.
```
Comment (zimulala) [inline plan/decorrelate.go]: s/decorrelate/Decorrelate

File: plan/decorrelate.go [hunk 5]
```diff
@@ -71,15 +74,16 @@ func decorrelate(p LogicalPlan) LogicalPlan {
...
 			innerPlan.SetParents(apply)
-			return decorrelate(p)
+			return s.optimize(p, nil, nil)
 		}
 		// TODO: Deal with aggregation and projection.
 	}
 	newChildren := make([]Plan, 0, len(p.GetChildren()))
 	for _, child := range p.GetChildren() {
-		newChildren = append(newChildren, decorrelate(child.(LogicalPlan)))
+		np, _ := s.optimize(child.(LogicalPlan), nil, nil)
```
Comment (zimulala) [inline plan/decorrelate.go]: Why not check this error?

Comment (hanfei1991) [inline plan/decorrelate.go]: the returned err always be nil.

**Comment by shenli** [comment]
LGTM
Please resolve the conflict.

**Comment by shenli** [comment]
@zimulala @coocood PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**